### ISSUE TITLE
[Enhancement] Add Yield Classification Model (Low/Medium/High) with Confidence Score to Crop Yield Prediction

### ISSUE THREAD

**Issue by ta2012** [PRIYANSHU2026/DataSentience-AIML#238]
## 📈 Feature Overview

I would like to contribute a **classification-based version** of the Crop Yield Prediction model in:

`src/Agricultural Solutions/Crop Yield Prediction/`

This model will categorize crop yields into the following classes:

| Label   | Yield Range (hg/ha) | Code |
|---------|---------------------|------|
| 🔴 Low     | Less than 20          | 0    |
| 🟡 Medium  | 20 to 30              | 1    |
| 🟢 High    | Greater than 30       | 2    |

---

## 🔧 Proposed Implementation

- Create a new training script `crop_yield_classifier.py`
- Use pandas to bin `hg/ha_yield` into 3 custom classes
- Train a `RandomForestClassifier`
- Save model as `classifier_model.pkl`
- Add a function `classify_yield(features)` inside `predict.py` which:
  - Loads the classifier
  - Predicts the class (Low / Medium / High)
  - Returns the top class with a confidence score using `predict_proba()`

---

## 🧪 Sample Output

```json
{
  "predicted_class": "Medium",
  "confidence": 82.6
}
```

## 🙋‍♂️ My Role
I will:

- Build and train the classification model

-Write clean, testable prediction logic using .predict_proba()

-Keep the original regression model intact

-Return output in clean JSON format for future frontend/API integration

## 📂 Folder Path

src/Agricultural Solutions/Crop Yield Prediction/
## ✅ Request
Kindly assign this issue to me under GSSoC 2025.
I am ready to implement this feature immediately. Looking forward to your approval.

Thank you!
Tanmay Bhardwaj
GSSoC Contributor

**Comment by ta2012** [comment]
Hey,  
I’m excited to contribute this enhancement under GSSoC 2025!  
Please assign this issue to me — I’m ready to start right away 🚀  
Thanks!

**Comment by MahraibFatima** [comment]
@ta2012 you can work on this, now.
best of luck.

**Comment by MahraibFatima** [comment]
@ta2012 what's the update?

**Comment by ta2012** [comment]
Hi Mahraib,

Thanks for checking in!

I’ve spent some time analyzing the current data distribution and exploring
potential approaches for the classification task. I'm currently evaluating
two strategies:

1. Regression-to-Classification Mapping:
Using the existing regression model outputs, I’m defining dynamic threshold
bands (based on quantiles or domain-specific yield ranges) to classify
predictions into Low, Medium, and High yield categories. This ensures
continuity with existing models while enabling confidence-based
classification.


2. Direct Multi-Class Classification:
I'm also experimenting with a supervised multi-class model (Random Forest
and XGBoost so far) trained directly on labeled yield categories. This
approach may simplify confidence score extraction via softmax probabilities
or calibrated classifiers.



To support the confidence score requirement, I’m preparing to integrate
either:

Probability calibration (Platt Scaling / Isotonic Regression) for crisp
confidence estimates, or

Ensemble variance tracking if we stick to regression-first approaches.


I'll finalize the approach based on comparative F1/accuracy vs.
interpretability tradeoffs. Will push a structured notebook or draft
pipeline to the repo shortly.

Best regards,
Tanmay (ta2012)

**Comment by MahraibFatima** [comment]
that's great, you're learning while contributing. ✨

**Comment by MahraibFatima** [comment]
@ta2012 it's too much late.
just one more day and I'll unassign you from this issue.
